### ISSUE TITLE
HMR breaks page in presence of (dynamic) cyclic imports

### ISSUE THREAD

**Issue by charlag** [PepsRyuu/nollup#175]
Hi!
We run into this case while debugging issues with HMR. I think it should be possible to solve without live bindings because it works on first load.

Here's example branch:
https://github.com/tutao/nollup-bug/tree/cyclic-hmr

Basically root dynamically imports another module and this module in turn imports root (but statically). It loads but if you change util module it fails with exhausting call stack.
Unfortunately I didn't find a solution yet but would be nice to have. Such code organization is obviously flawed but it's a hard reality sometimes.

**Comment by PepsRyuu** [comment]
Thanks for the report! Won't be able to look at this for a couple of days unfortunately, but I promise I'll get back to you. 🙂

**Comment by charlag** [comment]
Don't worry, we solved our cyclies for now, you work is greatly appreciated!

**Comment by PepsRyuu** [comment]
Thanks for the repro and apologies for only getting to this now. This seems to be related to dynamic imports not checking if a module is already resolved before trying to load it. I have a large refactor upcoming that I want to finalise first because the solution to this will be heavily impacted by it. 

---------

This is just for my own reference here for this issue:

```
module.__impl(localRequire, module, __nollup__global__);
module.__resolved = true;
```

Swapping these two lines seems to resolve the issue, needs further testing with proper test cases.

**Comment by charlag** [comment]
Thanks a lot for looking at this, it is really appreciated. While it doesn't disturb is too much right now it would be great to fix it.
This thing about check sounds about right, does it mean that other dynamic imports are also loaded (and instanciated) multiple times?

**Comment by PepsRyuu** [comment]
Yeah it seems likely dynamic imports are executed multiple times right now, which certainly doesn't seem right. Will need to define a bunch of test cases for this and look at it in greater detail, so might be some time before this is fully addressed. I'm hoping to have a big refactor done very soon (was delayed because of an emergency last week), and once that's done, I can have a proper look at this as the refactor slightly changes how dynamic imports work.

**Comment by charlag** [comment]
Let me know if I can help you anyhow and thanks a lot for checking this, take care!

**Comment by PepsRyuu** [comment]
If you can give this branch a try, I'd greatly appreciate it! https://github.com/PepsRyuu/nollup/pull/173

**Comment by charlag** [comment]
I will try it as soon as I am able to, thanks!

**Comment by charlag** [comment]
I tested it on the example I provided and on [Tutanota](github.com/tutao/tutanota) (which should have no such cycles anymore anyway) and both seem to work well!

**Comment by PepsRyuu** [comment]
Thanks! I'll look into releasing later this week hopefully. :)

**Comment by PepsRyuu** [comment]
Released in ```0.15.0```
